File: src/infrastructure/tools/execute/execute-script.ts
/**
 * execute_script — 创建脚本文件并执行，可选执行后立即删除
 *
 * 与 `execute_command` 的分工：后者执行一条内联 shell 命令（多行脚本在内联时转义 / 引号很痛苦），本工
 * 具适合「需要一段较长的脚本」—— 先落盘成文件再执行，执行完默认删除，避免污染工作目录。
 *
 * 审批：独立门禁 `script.execute`（与终端命令风险分类无关）；申请不使用沙盒（`sandbox:"off"`）时另过
 * 「沙盒脱壳·脚本执行」门禁（取更严格者）；命中「忽略沙盒命令」规则（设置 → 安全）时免脱壳审批并强制
 * 无沙盒执行。写文件走 `securityService.resolveSafePath(mode='w')`，超范围直接报错。
 *
 * ⚠️ 规则语义与 Rust 原生路径（`execute_script.rs`）对齐，匹配对象是**运行命令**（不是脚本正文）。
 */
import * as tauriFs from '@tauri-apps/plugin-fs'
import { invoke } from '@tauri-apps/api/core'

import { toolRegistry } from '@/domain/tools'
import {
  UserInteractionRequired,
  type ToolContext,
  type ToolExecutor,
  type ToolResult,
} from '@/domain/tools/types'
import { t, tpl } from '@/ui/i18n'
import { securityService } from '@/services/security-service'
import {
  PERM_SANDBOX_SCRIPT,
  PERM_SCRIPT,
  permissionLabel,
  resolveCommandDecision,
} from '@/domain/permission'
import { settingsState } from '@/ui/store'
import { track } from '@/utils/telemetry'
import {
  SANDBOX_BYPASS_HINT,
  SANDBOX_RULE_BYPASS_HINT,
  classifyCommand,
  detectPlatform,
  getRiskInfo,
  registerPendingApproval,
  runCommand,
} from './common'

// 工具描述（含三平台变体）已收敛到权威源（机制 C）：
// src-tauri/virlen-core/src/agent/tool_defs/definitions.json —— 见 docs/rust-engine.md §12。

toolRegistry.register(
    'execute_script',
    (async (
    args: Record<string, any>,
    ctx: ToolContext,
  ): Promise<ToolResult | UserInteractionRequired> => {
    if (!tauriFs)
      throw '[execute_script] Error: not running in a Tauri environment'

    const filePath = args.file_path as string
    const content = String(args.file_content ?? '')
    const command = args.command as string
    const tips = args.tips as string | undefined
    const endDelFile = args.end_del_file !== false // 默认 true
    const rawTimeout = Number(args.timeout ?? 30)
    let timeout = Number.isFinite(rawTimeout) ? rawTimeout : 30
    if (timeout < 0) timeout = 30
    if (timeout > 300) timeout = 300
    const timeoutMs = timeout * 1000

    // ⚠️ 以下报错为模型侧文案：固定英文，与 Rust `execute_script.rs` 逐字对齐（铁律 1）。
    if (!filePath) throw 'Missing required parameter: "file_path"'
    if (!command) throw 'Missing required parameter: "command"'

    const fullPath = await securityService.resolveSafePath(
      filePath,
      'w',
      ctx.sessionId,
    )
    const cwd = await securityService.getWorkspace(ctx.sessionId)

    // sandbox: 'off' → 申请「不使用沙盒」执行脚本（与 execute_command 同语义）。
    const aiRequestedBypass = ['off', 'none'].includes(
      String(args.sandbox ?? '').toLowerCase(),
    )
    const sandboxMode = settingsState.value.sandboxMode ?? 'on'
    // 只读模式禁止 AI 显式申请绕过沙盒（与 Rust 原生路径一致）。
    // ⚠️ 但只针对 AI 的显式申请：命中「忽略沙盒命令」规则时只读模式静默忽略规则（脚本继续走沙盒）。
    if (aiRequestedBypass && sandboxMode === 'readonly') {
      throw new Error(
        'The sandbox is in read-only mode, so bypassing it to run a script is not allowed; switch the sandbox mode in settings first (or use a regular terminal)',
      )
    }
    // 「忽略沙盒命令」规则（设置 → 安全）：命中即**免脱壳审批 + 强制无沙盒执行**。
    // 匹配对象是**运行命令**（如 node temp/run.js），不是脚本正文（与 Rust 侧一致）。
    const ruleHit =
      sandboxMode === 'on'
        ? await securityService.matchSandboxIgnoreRule(command)
        : null
    const bypassSandbox = aiRequestedBypass || !!ruleHit
    if (ruleHit) {
      // 留痕（只记工具名 / 原因，不记命令正文与规则名，遵循 §9）
      track('tool.sandbox.bypass', {
        tool_name: 'execute_script',
        status: 'auto_rule',
      })
    }

    // 目标文件已存在则驳回，避免覆盖既有文件
    if (await tauriFs.exists(fullPath).catch(() => false)) {
      throw `Error: the script file already exists; refusing to overwrite it — ${fullPath}`
    }

    // 写文件 → 执行命令 → （可选）删除文件
    const writeRunDelete = async (): Promise<ToolResult> => {
      await writeScriptFile(fullPath, content)

      ctx.write(`> ${command}\n`)

      try {
        const result = await runCommand(
          command,
          cwd,
          timeoutMs,
          ctx,
          'execute_script',
          { bypassSandbox },
        )
        if (endDelFile) {
          const note = await deleteScriptFile(fullPath)
          return {
            content: result.content + '\n' + note.text,
            uiData: {
              ...(result.uiData ?? {}),
              // note：模型侧 + 旧消息回退用的英文文本；其余三个字段供 UI 本地化渲染
              note: note.text,
              noteKind: note.kind,
              notePath: note.path,
              ...(note.error ? { noteError: note.error } : {}),
            },
          }
        }
        return result
      } catch (e) {
        // 执行失败/超时/取消：也照常清理脚本，避免残留
        if (endDelFile) {
          await deleteScriptFile(fullPath).catch(() => {})
        }
        throw e
      }
    }

    // 权限：脚本执行独立门禁（script.execute，默认每次弹窗；与命令风险分类无关）
    const base = await securityService.getPermissionDecision(PERM_SCRIPT)
    // 申请绕过沙盒且沙盒启用 → 额外过「沙盒脱壳·脚本执行」权限（与脚本权限取更严格者）
    const configuredEscape =
      bypassSandbox && sandboxMode !== 'off'
        ? await securityService.getPermissionDecision(PERM_SANDBOX_SCRIPT)
        : undefined
    // 命中规则 → 用户已用规则预先授权脱壳（ask 视作 allow）；⚠️ deny 优先
    const escapeDecision =
      ruleHit && configuredEscape === 'ask' ? 'allow' : configuredEscape
    const decision = resolveCommandDecision(base, { escapeDecision })
    if (decision === 'deny') {
      // 禁止：不执行、不弹窗，返回拒绝文本给模型（标明是哪个权限拦下的）
      const deniedPerm =
        escapeDecision === 'deny' ? PERM_SANDBOX_SCRIPT : PERM_SCRIPT
      // 只报**权限 name**（与设置页一一对应的稳定 key）：语言无关，
      // 且与 Rust 原生实现逐字对齐
      throw new Error(
        `Operation denied by the permission settings: ${deniedPerm}`,
      )
    }
    if (decision === 'allow') {
      return writeRunDelete()
    }

    // ask → 弹窗审批：脚本内容无法静态分析，一律交由用户确认。
    const risk = classifyCommand(command)
    const info = getRiskInfo(risk)
    const baseHint =
      info.hint || t('此操作会创建并执行脚本文件，请确认是否允许')
    // 追加沙盒脱壳警告（让用户看到后果）：命中规则时说明「为什么没申请也脱壳了」
    const hint = ruleHit
      ? [baseHint, tpl(SANDBOX_RULE_BYPASS_HINT, { rule: ruleHit.name })]
          .filter(Boolean)
          .join('\n')
      : bypassSandbox
        ? [baseHint, t(SANDBOX_BYPASS_HINT)].filter(Boolean).join('\n')
        : baseHint
    // 触发本次确认的权限（同上：仅因沙盒脱壳时展示脱壳权限）
    const shownPerm =
      bypassSandbox && base === 'allow' && escapeDecision === 'ask'
        ? PERM_SANDBOX_SCRIPT
        : PERM_SCRIPT
    const approvalId = registerPendingApproval({
      sessionId: ctx.sessionId,
      toolCallId: ctx.toolCallId,
      run: writeRunDelete,
    })
    const payload: Record<string, any> = {
      approvalId,
      // 通用授权字段（弹窗展示）
      permName: shownPerm,
      title: t(permissionLabel(shownPerm)),
      subTitle: tips,
      // ⚠️ 正文展示脚本内容（用户据此判断是否放行），运行命令放在 command 作说明
      desc: content,
      command,
      hint,
      risk,
    }
    if (bypassSandbox) payload.sandboxBypass = true
    return new UserInteractionRequired('confirm_command', payload)
  }) as ToolExecutor,
    t('执行脚本'),
)

/** Windows 上会按「系统 ANSI 代码页」解析无 BOM 脚本的 shell 扩展名（PowerShell 脚本） */
const PS_SCRIPT_EXT = /\.ps(m)?1$/i

/**
 * 给脚本内容加 UTF-8 BOM（幂等）—— 与 Rust 侧 `with_script_bom` 等价。
 *
 * 必须加：PowerShell 5.1 读无 BOM 的 .ps1 时不猜 UTF-8，而按系统 ANSI 代码页
 * （中文系统 CP936/GBK）解析，中文字面量在解析阶段就已乱码（"脚本" → "鑴氭湰"），
 * 之后设 `[Console]::OutputEncoding` 也还原不回来（问题在输入端）。带 BOM 即按 UTF-8 解析。
 *
 * 只对 Windows 的 .ps1/.psm1 生效：.sh 加 BOM 会让 shebang 失效，.js/.py 虽能容忍但没必要。
 *
 * @param filePath 脚本完整路径（按扩展名判定是否需要 BOM）
 * @param platform Rust `os_platform` 返回值（windows / macos / linux）
 */
export function applyScriptBom(
  filePath: string,
  content: string,
  platform: string,
): string {
  if (platform !== 'windows' || !PS_SCRIPT_EXT.test(filePath)) return content
  return content.startsWith('\uFEFF') ? content : '\uFEFF' + content
}

/** 写入脚本文件（自动创建父目录） */
async function writeScriptFile(
  fullPath: string,
  content: string,
): Promise<void> {
  try {
    const normalizedPath = fullPath.replace(/\\/g, '/')
    const parent = normalizedPath.substring(0, normalizedPath.lastIndexOf('/'))
    if (parent) {
      await tauriFs.mkdir(parent, { recursive: true }).catch(() => { })
    }
    // PowerShell 脚本需要 UTF-8 BOM，否则 5.1 会按 GBK 解析导致中文乱码（见 applyScriptBom）
    await tauriFs.writeTextFile(
      fullPath,
      applyScriptBom(fullPath, content, await detectPlatform()),
    )
  } catch (e: any) {
    throw `Error: failed to write the script file — ${e?.message || String(e)}`
  }
}

/** 脚本删除结果：模型侧英文文本 + 供 UI 按界面语言渲染的结构化字段。 */
interface ScriptDeleteNote {
  /** 模型侧文本（与 Rust 侧 `delete_script_file` 逐字对齐） */
  text: string
  kind: 'deleted' | 'delete_failed'
  path: string
  error?: string
}

/** 删除脚本文件（移至回收站），返回模型侧文本 + 结构化字段（UI 侧本地化渲染） */
async function deleteScriptFile(fullPath: string): Promise<ScriptDeleteNote> {
  try {
    await invoke('move_to_trash', { path: fullPath })
    return {
      text: `🗑️ Script file deleted: ${fullPath}`,
      kind: 'deleted',
      path: fullPath,
    }
  } catch (e: any) {
    const error = e?.message || String(e)
    return {
      text: `⚠️ Failed to delete the script file: ${fullPath} — ${error}`,
      kind: 'delete_failed',
      path: fullPath,
      error,
    }
  }
}

File: src-tauri/virlen-core/src/agent/native_tools/execute/execute_script.rs
//! `execute_script` 工具（原生）— 创建脚本文件并执行，可选执行后立即删除。
//!
//! 流程：沙盒校验路径 → （按 script.execute 权限三态审批）→ 写脚本 → 执行 → 按 end_del_file 删除。

use crate::agent::bridge::BridgeInteractionResult;
use crate::agent::native_tools::common::{arg_bool, arg_i64, arg_str, resolve_safe_path};
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use crate::file_ops;
use serde_json::{json, Value};

use super::common::{
    apply_rule_clearance, match_sandbox_ignore_rule, classify_command, command_decision,
    permission_label, resolve_decision, risk_info, run_command_native, sandbox_mode,
    with_bypass_hint, with_rule_hint, PermissionDecision, SandboxMode, PERM_SANDBOX_SCRIPT,
    PERM_SCRIPT,
};

/// 执行脚本工具（原生）— 创建脚本文件并执行，可选执行后立即删除。
///
/// 流程：沙盒校验路径 → （按 script.execute 权限三态审批）→ 写脚本 → 执行 → 按 end_del_file 删除。
pub(crate) async fn execute_script_tool(
    ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    let file_path = arg_str(args, "file_path").unwrap_or_default();
    if file_path.trim().is_empty() {
        return Err("Missing required parameter: \"file_path\"".to_string());
    }
    let content = arg_str(args, "file_content").unwrap_or_default();
    let cmd_str = arg_str(args, "command").unwrap_or_default();
    if cmd_str.trim().is_empty() {
        return Err("Missing required parameter: \"command\"".to_string());
    }
    let end_del_file = arg_bool(args, "end_del_file").unwrap_or(true);
    let mut timeout = arg_i64(args, "timeout").unwrap_or(30);
    if timeout < 0 {
        timeout = 30;
    }
    if timeout > 300 {
        timeout = 300;
    }

    // sandbox:"off" → 申请「不使用沙盒」执行脚本（与 execute_command 同语义）。
    // ⚠️ 只读模式直接拒绝，否则只读保护会被绕过。
    let ai_requested_bypass = matches!(
        arg_str(args, "sandbox")
            .unwrap_or_default()
            .to_ascii_lowercase()
            .as_str(),
        "off" | "none"
    );
    if ai_requested_bypass && sandbox_mode(ctx) == SandboxMode::Readonly {
        return Err(
            "The sandbox is in read-only mode, so bypassing it to run a script is not allowed; switch the sandbox mode in settings first (or use a regular terminal)"
                .to_string(),
        );
    }

    // 沙盒校验（写权限），非法路径快速失败
    let full_path = resolve_safe_path(&file_path, "w", ctx.security)?;

    // 目标文件已存在则驳回，避免覆盖既有文件
    if std::path::Path::new(&full_path).exists() {
        return Err(format!(
            "Error: the script file already exists; refusing to overwrite it — {}",
            full_path
        ));
    }

    // 「忽略沙盒命令」规则（设置 → 安全）：与 execute_command 同语义 —— 命中即免脱壳审批并强制无沙盒执行
    // （AI 没传 sandbox:"off" 也生效），匹配对象是运行命令（不是脚本正文，见 `common::rules` 模块头注释）。
    // 放在「脚本已存在」快速失败之后：不值得为一条必然报错的调用多判一次规则。
    // ⚠️ 只在沙盒启用时判定：off 时无沙盒可脱；readonly 时脱壳被禁止（规则静默忽略）。
    let rule_hit = if sandbox_mode(ctx) == SandboxMode::On {
        match_sandbox_ignore_rule(ctx, &cmd_str).await
    } else {
        None
    };
    if rule_hit.is_some() {
        // 留痕（只记工具名 / 原因，不记命令正文与规则名，遵循 §9）
        crate::telemetry::track(
            "tool.sandbox.bypass",
            json!({ "tool_name": "execute_script", "status": "auto_rule" }),
        );
    }
    // 实际是否以「不使用沙盒」方式执行：AI 显式申请 ∪ 命中规则
    let bypass_sandbox = ai_requested_bypass || rule_hit.is_some();

    // 权限：脚本执行独立门禁（script.execute，默认每次弹窗；与命令风险分类无关）。
    // permissions 表优先，回退 legacy approval_mode（兼容老客户端 / 测试）。
    let base = command_decision(
        &ctx.security.permissions,
        &ctx.security.approval_mode,
        PERM_SCRIPT,
        "safe",
    );
    // 申请绕过沙盒且沙盒启用（readonly 已在上方直接拒绝）→ 额外过「沙盒脱壳·脚本执行」门禁
    // （与脚本权限**取更严格者**，默认 ask）。沙盒模式 off 时无沙盒可脱，不参与门禁。
    // 脱壳权限无 legacy 对应项 → approval_mode 传空串，只用权限表 / 注册表默认（ask）。
    let escape_decision = if bypass_sandbox && sandbox_mode(ctx) != SandboxMode::Off {
        let configured = command_decision(
            &ctx.security.permissions,
            "",
            PERM_SANDBOX_SCRIPT,
            "safe",
        );
        // 命中「忽略沙盒命令」规则 → 用户已用规则预先授权脱壳（ask 视作 allow）；
        // ⚠️ deny 仍优先：规则不能推翻显式禁止
        Some(if rule_hit.is_some() {
            apply_rule_clearance(configured)
        } else {
            configured
        })
    } else {
        None
    };
    match resolve_decision(base, escape_decision, false) {
        PermissionDecision::Deny => {
            let denied = if escape_decision == Some(PermissionDecision::Deny) {
                PERM_SANDBOX_SCRIPT
            } else {
                PERM_SCRIPT
            };
            // 只报**权限 name**（稳定 key，与设置页一一对应）：语言无关，
            // 且与 TS 执行器（`tools/execute/execute-script.ts`）逐字对齐（铁律 1）。
            return Err(format!(
                "Operation denied by the permission settings: {}",
                denied
            ));
        }
        PermissionDecision::Allow => {
            return finalize_script_run(
                ctx,
                &cmd_str,
                timeout,
                &full_path,
                &content,
                end_del_file,
                bypass_sandbox,
            )
            .await;
        }
        PermissionDecision::Ask => {}
    }

    // ask → 弹窗审批：脚本内容无法静态分析（命令名可能只是 node/python，
    // 但脚本内部可能是任意代码），一律交由用户确认。
    let risk = classify_command(&cmd_str);
    let (_base_label, base_hint) = risk_info(risk);
    let hint = if base_hint.is_empty() {
        "此操作会创建并执行脚本文件，请确认是否允许".to_string()
    } else {
        base_hint
    };
    // 追加沙盒脱壳警告（让用户看到后果）：命中规则时说明「为什么没申请也脱壳了」
    let hint = match &rule_hit {
        Some(rule_name) => with_rule_hint(&hint, rule_name),
        None if bypass_sandbox => with_bypass_hint(&hint),
        None => hint,
    };
    // 触发本次确认的权限（同上：仅因沙盒脱壳时展示脱壳权限）
    let shown_perm = if bypass_sandbox
        && base == PermissionDecision::Allow
        && escape_decision == Some(PermissionDecision::Ask)
    {
        PERM_SANDBOX_SCRIPT
    } else {
        PERM_SCRIPT
    };
    let tips = arg_str(args, "tips").unwrap_or_default();
    let mut data = json!({
        // 通用授权字段（弹窗展示）
        "permName": shown_perm,
        "title": permission_label(shown_perm),
        "subTitle": tips,
        // 正文展示脚本内容（用户据此判断是否放行），运行命令放在 command 作说明
        "desc": content,
        "command": cmd_str,
        "hint": hint,
        "risk": risk,
    });
    if let Value::Object(map) = &mut data {
        map.insert(
            "toolCallId".into(),
            Value::String(ctx.tool_call_id.to_string()),
        );
        if bypass_sandbox {
            // 供弹窗高亮 / 埋点识别（文案已在 hint 里）
            map.insert("sandboxBypass".into(), Value::Bool(true));
        }
    }

    let payload = ctx
        .bridge
        .request_user_interaction(ctx.sink, ctx.session_id, "confirm_command_native", data)
        .await
        .map_err(|e| format!("error: {}", e))?;

    match BridgeInteractionResult::parse(&payload) {
        BridgeInteractionResult::Value {
            content: interaction_msg,
            ..
        } => {
            let normalized = interaction_msg.trim().to_lowercase();
            if normalized == "approved" || normalized == "允许" || interaction_msg == "ok" {
                // 用户允许 → 写脚本 + 执行 + 按需删除
                return finalize_script_run(
                    ctx,
                    &cmd_str,
                    timeout,
                    &full_path,
                    &content,
                    end_del_file,
                    bypass_sandbox,
                )
                .await;
            }
            // 未实际执行（拒绝/其他）→ 未落盘，无需清理。
            // 必须走 Error 通道：脚本一行都没跑，UI 不能显示成绿色「成功」。
            Ok(NativeToolOutcome::error(interaction_msg))
        }
        BridgeInteractionResult::Error { content, ui_data } => {
            Ok(NativeToolOutcome::Error { content, ui_data })
        }
        BridgeInteractionResult::Shelved => Ok(NativeToolOutcome::Shelved),
        // 用户拒绝授权 / Esc 取消 → 脚本未落盘、未执行 → 同样按失败回报
        // （与 execute_command 原生路径、JS 桥路径、TS 引擎保持一致）
        BridgeInteractionResult::Cancelled => {
            Ok(NativeToolOutcome::error("[User cancelled]"))
        }
    }
}

/// 目标文件是否为 PowerShell 脚本（Windows PowerShell 5.1 会按系统 ANSI 代码页解析无 BOM 的这类文件）。
fn is_powershell_script(path: &str) -> bool {
    let lower = path.to_ascii_lowercase();
    lower.ends_with(".ps1") || lower.ends_with(".psm1")
}

/// 给脚本内容补 UTF-8 BOM（幂等）—— 与 JS 侧 `applyScriptBom` 等价。
///
/// 必须加：PowerShell 5.1 读无 BOM 的 .ps1 时按系统 ANSI 代码页解析，中文字面量在解析阶段就变乱码
/// （"脚本" → "鑴氭湰"），之后再设 `[Console]::OutputEncoding` 也还原不回来。只对 Windows 的 .ps1/.psm1
/// 生效（.sh 加 BOM 会让 shebang 失效）。
fn with_script_bom(path: &str, content: &str, is_windows: bool) -> String {
    if is_windows && is_powershell_script(path) && !content.starts_with('\u{FEFF}') {
        format!("\u{FEFF}{}", content)
    } else {
        content.to_string()
    }
}

/// 脚本执行收尾：写脚本 → 执行命令 → 按 end_del_file 删除脚本。
///
/// `bypass_sandbox`：`sandbox:"off"` 申请（已在上方过「沙盒脱壳」权限门禁）→ 裸跑。
async fn finalize_script_run(
    ctx: &NativeToolCtx<'_>,
    cmd_str: &str,
    timeout_secs: i64,
    full_path: &str,
    content: &str,
    end_del_file: bool,
    bypass_sandbox: bool,
) -> Result<NativeToolOutcome, String> {
    // 1. 写脚本文件（自动创建父目录）
    let full_path_c = full_path.to_string();
    // 落盘内容可能与入参不同：Windows 的 .ps1/.psm1 需补 UTF-8 BOM（见 with_script_bom）
    let content_c = with_script_bom(full_path, content, cfg!(target_os = "windows"));
    let write_res =
        tokio::task::spawn_blocking(move || file_ops::write_file(&full_path_c, &content_c)).await;
    match write_res {
        Ok(Ok(_)) => {}
        Ok(Err(e)) => return Err(format!("Error: failed to write the script file — {}", e)),
        Err(e) => return Err(format!("Task join error: {}", e)),
    }

    // 2. 执行命令（bypass_sandbox 来自 sandbox:"off" 申请，已在上方过权限门禁）
    let outcome = run_command_native(ctx, cmd_str, timeout_secs, bypass_sandbox).await;

    // 3. 按需删除脚本（含失败/超时）
    if !end_del_file {
        return outcome;
    }
    let del_note = delete_script_file(full_path).await;
    match outcome {
        Ok(NativeToolOutcome::Value { content, ui_data }) => Ok(NativeToolOutcome::Value {
            content: format!("{}\n{}", content, del_note.text),
            ui_data: attach_note(ui_data, &del_note),
        }),
        Ok(NativeToolOutcome::Error { content, ui_data }) => Ok(NativeToolOutcome::Error {
            content: format!("{}\n{}", content, del_note.text),
            // 失败同样保留结构化字段（并在有 note 时补上）——UI 才能按界面语言渲染（L6）
            ui_data: attach_note(ui_data, &del_note),
        }),
        Err(e) => Err(format!("{}\n{}", e, del_note.text)),
        Ok(other) => Ok(other),
    }
}

/// 脚本删除结果：模型侧英文文本 + 供 UI 按界面语言渲染的结构化字段。
///
/// ⚠️ 与 JS 侧 `ScriptDeleteNote`（`tools/execute/execute-script.ts`）逐字对齐（铁律 1）：`text` 为模型侧
/// 文案，`kind` / `path` / `error` 是语言无关数据，由 `TerminalBlock` 重建展示文本（旧消息无这些字段 → 回退
/// `note` 文本）。
struct ScriptDeleteNote {
    text: String,
    kind: &'static str,
    path: String,
    error: Option<String>,
}

/// 把删除提示写入 ui_data（供 UI 展示 / 本地化渲染），保留原有字段。
fn attach_note(ui_data: Option<Value>, note: &ScriptDeleteNote) -> Option<Value> {
    let with_note = |mut map: serde_json::Map<String, Value>| {
        map.insert("note".into(), Value::String(note.text.clone()));
        map.insert("noteKind".into(), Value::String(note.kind.to_string()));
        map.insert("notePath".into(), Value::String(note.path.clone()));
        if let Some(err) = &note.error {
            map.insert("noteError".into(), Value::String(err.clone()));
        }
        map
    };
    match ui_data {
        Some(Value::Object(map)) => Some(Value::Object(with_note(map))),
        Some(other) => {
            let mut map = serde_json::Map::new();
            map.insert("data".into(), other);
            Some(Value::Object(with_note(map)))
        }
        None => Some(Value::Object(with_note(serde_json::Map::new()))),
    }
}

/// 删除脚本文件（移至回收站），返回模型侧文本 + 结构化字段。
async fn delete_script_file(full_path: &str) -> ScriptDeleteNote {
    let p = full_path.to_string();
    match tokio::task::spawn_blocking(move || trash::delete(&p)).await {
        Ok(Ok(_)) => ScriptDeleteNote {
            text: format!("🗑️ Script file deleted: {}", full_path),
            kind: "deleted",
            path: full_path.to_string(),
            error: None,
        },
        Ok(Err(e)) => script_delete_failure(full_path, e.to_string()),
        Err(e) => script_delete_failure(full_path, format!("Task join error: {}", e)),
    }
}

/// 删除失败的统一构造（`text` 必须与 TS 侧 `deleteScriptFile` 的失败分支逐字对齐）。
fn script_delete_failure(full_path: &str, error: String) -> ScriptDeleteNote {
    ScriptDeleteNote {
        text: format!(
            "⚠️ Failed to delete the script file: {} — {}",
            full_path, error
        ),
        kind: "delete_failed",
        path: full_path.to_string(),
        error: Some(error),
    }
}

#[cfg(test)]
mod tests {
    use super::{attach_note, is_powershell_script, script_delete_failure, with_script_bom, ScriptDeleteNote};
    use serde_json::json;

    #[test]
    fn script_delete_note_structure() {
        // 失败分支：文案与 TS 侧 `deleteScriptFile` 逐字对齐，且带结构化字段（UI 据此本地化）
        let note = script_delete_failure("C:/ws/run.js", "boom".to_string());
        assert_eq!(note.kind, "delete_failed");
        assert_eq!(
            note.text,
            "⚠️ Failed to delete the script file: C:/ws/run.js — boom"
        );
        assert_eq!(note.error.as_deref(), Some("boom"));
        assert_eq!(note.path, "C:/ws/run.js");
    }

    #[test]
    fn attach_note_preserves_existing_fields() {
        let note = ScriptDeleteNote {
            text: "note-text".to_string(),
            kind: "deleted",
            path: "C:/ws/run.js".to_string(),
            error: None,
        };
        // 原有 ui_data 字段保留，只追加 note*
        let ui = attach_note(Some(json!({ "stdout": "hi" })), &note).unwrap();
        assert_eq!(ui["stdout"], "hi");
        assert_eq!(ui["note"], "note-text");
        assert_eq!(ui["noteKind"], "deleted");
        assert_eq!(ui["notePath"], "C:/ws/run.js");
        assert!(ui.get("noteError").is_none());
        // 非对象 ui_data 也不丢（包进 data）
        let ui2 = attach_note(Some(json!("raw")), &note).unwrap();
        assert_eq!(ui2["data"], "raw");
        assert_eq!(ui2["noteKind"], "deleted");
        // 无 ui_data → 只带 note 字段
        let ui3 = attach_note(None, &note).unwrap();
        assert_eq!(ui3["noteKind"], "deleted");
        assert!(ui3.get("stdout").is_none());
    }

    #[test]
    fn powershell_script_ext_detection() {
        assert!(is_powershell_script("C:/ws/temp/run.ps1"));
        assert!(is_powershell_script("C:\\ws\\temp\\run.PSM1"));
        assert!(!is_powershell_script("C:/ws/temp/run.js"));
        assert!(!is_powershell_script("C:/ws/temp/run.sh"));
        assert!(!is_powershell_script("C:/ws/temp/ps1"));
        assert!(!is_powershell_script("C:/ws/temp/run.ps1.bak"));
    }

    #[test]
    fn bom_only_for_windows_powershell() {
        // Windows + .ps1 → 补 BOM（PowerShell 5.1 据此按 UTF-8 解析，中文才不乱码）
        let out = with_script_bom("C:/ws/run.ps1", "Write-Output \"脚本\"", true);
        assert!(out.starts_with('\u{FEFF}'));
        assert_eq!(&out[3..], "Write-Output \"脚本\"");
        // 非 Windows → 不加（macOS/Linux 的 pwsh 默认按 UTF-8 解析）
        assert_eq!(with_script_bom("C:/ws/run.ps1", "x", false), "x");
        // Windows 上的非 PowerShell 脚本 → 不加（.sh 加 BOM 会让 shebang 失效）
        assert_eq!(with_script_bom("C:/ws/run.sh", "x", true), "x");
        assert_eq!(with_script_bom("C:/ws/run.js", "x", true), "x");
    }

    #[test]
    fn bom_is_idempotent() {
        let once = with_script_bom("C:/ws/run.ps1", "abc", true);
        let twice = with_script_bom("C:/ws/run.ps1", &once, true);
        assert_eq!(once, twice);
    }
}
